### PR TITLE
Truncate long queries in error logs

### PR DESCRIPTION
**Pull request by demmer**
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/2794)

### CODE REVIEW

File: go/vt/vttablet/tabletserver/query_engine.go [hunk 1]
```diff
@@ -470,7 +470,7 @@ func (qe *QueryEngine) handleHTTPQueryPlans(response http.ResponseWriter, reques
 	response.Header().Set("Content-Type", "text/plain")
 	response.Write([]byte(fmt.Sprintf("Length: %d\n", len(keys))))
 	for _, v := range keys {
-		response.Write([]byte(fmt.Sprintf("%#v\n", utils.TruncateQuery(v))))
+		response.Write([]byte(fmt.Sprintf("%#v\n", utils.TruncateQuery(v, 512))))
```
Comment (sougou) [inline go/vt/vttablet/tabletserver/query_engine.go]: I'll be fine with using sqldb.SQLErrorTruncateLen everywhere uniformly.

**Comment by sougou** [comment]
@michael-berlin Do you have any thoughts? Change looks good to me, except for that one comment.

File: go/vt/vttablet/tabletserver/query_engine.go (hunk 1 above)
Comment (demmer) [inline go/vt/vttablet/tabletserver/query_engine.go]: My only thought on that is that they are somewhat orthogonal -- truncating in the UI isn't actually affecting errors so I'd want to name it something different.

**Comment by demmer** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[go/sqldb/sql_error.go, line 22 at r3](https://reviewable.io:443/reviews/youtube/vitess/2794#-KiHKodOuxqdZ-b8b_Wz:-KiJ0YbA0wNJwyR2aKKL:bbos4wg) ([raw file](https://github.com/youtube/vitess/blob/4ec08569843e91fb61ee10a4d56761a70a3a764c/go/sqldb/sql_error.go#L22)):*
<details><summary><i>Previously, mberlin-bot (Michael Berlin) wrote…</i></summary><blockquote>

It's also used in vtgate. Let's definitely make this a constant (ideally in go/vt/utils/utils.go) and not duplicate this magic value of 512.

Instead of a constant, we could also add a function for that.

```go
// TruncateQueryForUI is used when displaying queries on various Vitess status pages.
// By truncating the query to 512 characters, we avoid that loading the page takes too long.
func TruncateQueryForUI(query string) string {
  return TruncateQuery(query, 512)
}
```
</blockquote></details>

I had the same thought actually and did basically the same thing.

Will polish up the tests and make sure it covers the cases that have been biting us in production.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2794)*

**Comment by demmer** [comment]
Done

---

Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[go/sqldb/sql_error.go, line 10 at r3](https://reviewable.io:443/reviews/youtube/vitess/2794#-KiHL3YC8kngC9Cu8Wqn:-KiLT5B6k8TpUxHG6GHv:b-cxchxo) ([raw file](https://github.com/youtube/vitess/blob/4ec08569843e91fb61ee10a4d56761a70a3a764c/go/sqldb/sql_error.go#L10)):*
<details><summary><i>Previously, mberlin-bot (Michael Berlin) wrote…</i></summary><blockquote>

Go readability: This must go in a separate group.
</blockquote></details>

No longer relevant.

---

*[go/vt/vttablet/endtoend/misc_test.go, line 536 at r3](https://reviewable.io:443/reviews/youtube/vitess/2794#-KiHH3x6-G43Q0vt--dV:-KiJ3xZOOYYDMSRa27CR:b-w0are0) ([raw file](https://github.com/youtube/vitess/blob/4ec08569843e91fb61ee10a4d56761a70a3a764c/go/vt/vttablet/endtoend/misc_test.go#L536)):*
<details><summary><i>Previously, mberlin-bot (Michael Berlin) wrote…</i></summary><blockquote>

Go readability: Surrounding parentheses are not needed.

It looks like you're skipping the Git precommit hook. I thought it would complain about this? Can you please double check?
</blockquote></details>

I ended up moving around repositories to work with my VM and the symlinks were pointed at dead-ends.

However even after fixing them I don't see to be able to get one of these to fail. 

---

*[go/vt/vttablet/tabletserver/query_engine.go, line 473 at r3](https://reviewable.io:443/reviews/youtube/vitess/2794#-KiHGPXVQ1OL6XWlO6u0-r3-473:-KiLT03a2EgtcX41omkO:bgxmcff) ([raw file](https://github.com/youtube/vitess/blob/4ec08569843e91fb61ee10a4d56761a70a3a764c/go/vt/vttablet/tabletserver/query_engine.go#L473)):*
<details><summary><i>Previously, mberlin-bot (Michael Berlin) wrote…</i></summary><blockquote>

The 512 should be a constant because you're using it in several places and somebody may change it in one place and forget about the others.

In the name of the constant you could reflect that the limit is meant as default for UI purposes.

In general, I think it makes sense to have two different values for UI truncating and log truncating.

How about just adding a second flag (instead of a constant)? The default value for the second flag could be 512 then.
</blockquote></details>

I ended up adding two functions and separate command line args for each.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2794#-:-KiLT4712v8WZ2CqqdtF:bb6un1j)*

**Comment by demmer** [comment]
@sougou @michael-berlin When manually forcing various errors I realized my previous attempts didn't actually truncate as expected.

Reworked the implementation and verified this properly truncates errors from vttablet and vtgate.

**Comment by sougou** [comment]
I was thinking you could also use sqlparser.SplitTrailingComment (https://github.com/youtube/vitess/blob/master/go/vt/sqlparser/comments.go#L14), and then truncate the query part only.

**Comment by demmer** [comment]
@sougou That's a good idea... though perhaps for cleanliness I should move all the truncate functions into the `sqlparser` package as opposed to `sqldb`?

**Comment by sougou** [comment]
Yeah. That sounds like a good idea.

**Comment by demmer** [comment]
@sougou Updated as per your suggestion to preserve the comments.

File: go/vt/vttablet/endtoend/misc_test.go [hunk 2]
```diff
@@ -518,3 +520,55 @@ func TestDBAStatements(t *testing.T) {
 		t.Errorf("RowsAffected: %d, want 4", qr.RowsAffected)
 	}
 }
+
+func TestLongQueryErrorTruncation(t *testing.T) {
```
Comment (sougou) [inline go/vt/vttablet/endtoend/misc_test.go]: In all these tests, it may be simpler and more robust to do equality comparison with the exact output we want.

File: go/vt/sqlparser/truncate_query.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// TruncateForUI is used when displaying queries on various Vitess status pages
+// to keep the pages small enough to load and render properly
+func TruncateForUI(query string) string {
+	return truncateQuery(query, *TruncateUILen)
+}
+
+// TruncateForError is used when displaying queries as part of error logs
+// to avoid overwhelming logging systems with potentially long queries and
+// bind value data.
+func TruncateForError(query string) string {
```
Comment (sougou) [inline go/vt/sqlparser/truncate_query.go]: Rename to `TruncateForLog`?

File: go/vt/vtgate/vtgate_test.go [hunk 4]
```diff
@@ -2709,3 +2710,31 @@ func TestParseKeyspaceOptionalShard(t *testing.T) {
 		}
 	}
 }
+
+func TestErrorTruncation(t *testing.T) {
+	*sqlparser.TruncateErrLen = 30
+
+	query := map[string]interface{}{
```
Comment (sougou) [inline go/vt/vtgate/vtgate_test.go]: Like in the other comment, you can build a `wantVars` variable and perform an equality comparison using `reflect.DeepEqual`.

**Comment by demmer** [comment]
@sougou Updated with your suggestions.
